spec.md

```diff
@@ -3149,6 +3167,184 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 | Snapshot list mismatch | 3 INVALID_ARGUMENT | Besides the general cases, this code SHOULD also be used to indicate when plugin supporting CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT detects a mismatch in the `snapshot_ids`. | If a mismatch is detected in the `snapshot_ids`, caller SHOULD use different `snapshot_ids`. |
 | Volume group snapshot does not exist | 5 NOT_FOUND | Indicates that a volume group snapshot corresponding to the specified `group_snapshot_id` does not exist. | Caller MUST verify that the `group_snapshot_id` is correct and that the volume group snapshot is accessible and has not been deleted before retrying with exponential back off. |
 
+### Snapshot Metadata Service RPCs
+
+The Snapshot Metadata service is an optional service that is used to retrieve metadata on the allocated blocks of a single snapshot,
+or the changed blocks between a pair of snapshots of the same block volume.
+Retrieval of the data blocks of a snapshot is not addressed by this service and it
+is assumed that existing mechanisms to read snapshot data will suffice.
+
+#### Metadata Format
+
+Block volume data ranges are specified by a sequence of `BlockMetadata` tuples.
+Tuples in a sequence are in ascending order of `byte_offset`, with no overlap
+between adjacent tuples.
+
+```protobuf
+
+// BlockMetadata specifies a data range.
+message BlockMetadata {
+  // This is the 0 based byte position in the volume or snapshot,
+  // measured from the start of the object.
+  uint64 byte_offset = 1;
+
+  // This is the size of the data range.
+  uint64 size_bytes = 2;
+}
+
+```
+
+The `BlockMetadataType` enumerated type describes how the `size_bytes` fields of `BlockMetadata`
+tuples may vary in a sequence. There are two prevalent styles in which data ranges are described:
+
+- The **block-based** style requires a fixed value for the `size_bytes` field in all tuples in a given sequence.
+- The **extent-based** style does not constrain the value of the `size_bytes` field in a sequence.
+
+The SnapshotMetadata service permits either style at the discretion of the
+plugin as long as the style does not change mid-stream in any given RPC.
+
+```protobuf
+enum BlockMetadataType {
+  // The FIXED_LENGTH value indicates that data ranges are
+  // returned in fixed size blocks.
+  FIXED_LENGTH=0;
+
+  // The VARIABLE_LENGTH value indicates that data ranges
+  // are returned in potentially variable sized extents.
+  VARIABLE_LENGTH=1;
+}
+
+```
+
+
+#### GetAllocated RPC
+
+The plugin must implement this RPC if it provides the SnapshotMetadata service.
+
+```protobuf
+// The GetAllocatedRequest message is used to solicit metadata on the
+// allocated blocks of a snapshot: i.e. this identifies the data ranges
+// that have valid data as they were the target of some previous write
+// operation on the volume.
+message GetAllocatedRequest {
```

Comment:
I think that makes sense. Many providers might be using this for volume operations @carlbraganza wdyt?
We also might want to add a way to pass the provider-specific params just like the way it's done with `CreateVolumeRequest`?

```
  // Plugin specific creation-time parameters passed in as opaque 
  // key-value pairs. This field is OPTIONAL. The Plugin is responsible 
  // for parsing and validating these parameters. COs will treat 
  // these as opaque.
  map<string, string> parameters = 4;
```

Reply:
Parameters are probably not required, as the CSI-driver should be able to fetch those based on the `snapshot_id`.

Reply:
@nixpanic In K8s, does the secret data come from the `csi.storage.k8s.io/provisioner-secret-[name|namespace]` fields of the `VolumeSnapshotClass.Spec.Parameters`?

Reply:
> Ceph-CSI would like to implement this functionality. In order to communicate with the Ceph cluster, the requests (this one and `GetDeltaRequest`) need to provide credentials like the `secrets` as in `CreateVolumeRequest`:
> 
> ```
> map<string, string> secrets = 5 [(csi_secret) = true];
> ```

why aren't backend secrets configured at the plugin level? why do they need to be part of the RPC here? is this an exercise in future-proofing?

Reply:
> why aren't backend secrets configured at the plugin level? why do they need to be part of the RPC here? is this an exercise in future-proofing?

Ceph-CSI for example has a single instance for the CSI-drivers, but (in Kubernetes) the StorageClass points to a particular cluster. Different StorageClasses can point to different clusters, each with their own secrets (also linked in the StorageClass). Because of the stateless nature, there is no guarantee that secrets can be (temporarily) stored for plugin before APIs of this new SnapshotMetadata service are called.

Reply:
> @nixpanic In K8s, do the secret data come from either of:
> 
>     * `csi.storage.k8s.io/snapshotter-secret-[name|namespace]` entries of the `VolumeSnapshotClass.Parameters`
> 
>     * `csi.storage.k8s.io/provisioner-secret-[name|namespace]` entries of a `StorageClass.Parameters`

Yes, that is right. The kubernetes-csi/external-snapshotter resolves those secrets and passes the contents in the CSI procedures.

Reply:
@jdef, I would say yes, this is an exercise in future proofing. Every other CSI RPC that we've added without a secrets field has later needed a secrets field to be added. My stance is that the default should be to require the CO to provide pre-RPC secrets and to figure out what those look like, unless we can prove it doesn't make sense to do so.

The underlying problem are the few CSI drivers that use per-volume secrets which the CO is responsible for managing. Those drivers find it impossible to implement new features until a secrets field is added and the CO is updated to provide the appropriate per-volume secrets.